src/rust.rs

```diff
@@ -463,6 +464,185 @@ macro_rules! rooted {
+        *self.stackTop = self.down;
+    }
+}
+
+// This structure reimplements a C++ class that uses virtual dispatch, so
+// use C layout to guarantee that vftable in CustomAutoRooter is in right place.
+#[repr(C)]
+pub struct CustomAutoRooter<T> {
+    pub _base: jsapi::CustomAutoRooter,
+    pub data: Option<T>
+}
+
+impl<T> CustomAutoRooter<T> {
+    pub unsafe fn add_to_root_stack(&mut self, cx: *mut JSContext) {
+        self._base._base.add_to_root_stack(cx);
+    }
+    pub unsafe fn remove_from_root_stack(&mut self) {
+        self._base._base.remove_from_root_stack();
+    }
+}
+
+use jsapi::_vftable_CustomAutoRooter as CustomAutoRooterVFTable;
+/// CustomAutoRooters use dynamic dispatch on the C++ side for custom tracing,
+/// so provide trace logic via vftable when creating object on Rust side.
+unsafe trait CustomAutoTraceable: Sized {
+    const vftable: CustomAutoRooterVFTable = CustomAutoRooterVFTable {
+        trace: Self::trace,
+    };
+
+    unsafe extern "C" fn trace(this: *mut ::std::os::raw::c_void, trc: *mut JSTracer) {
+        let this = this as *const Self;
+        let this = this.as_ref().unwrap();
+        Self::do_trace(this, trc);
+    }
+
+    /// Used by `CustomAutoTraceable` implementee to trace its contents.
+    /// Corresponds to virtual `trace` call in CustomAutoRooter superclasses (C++).
+    fn do_trace(&self, trc: *mut JSTracer);
+}
+
+unsafe impl<T: CustomTrace> CustomAutoTraceable for CustomAutoRooter<T> {
+    fn do_trace(&self, trc: *mut JSTracer) {
+        self.data.as_ref().unwrap().trace(trc);
+    }
+}
+
+impl<T: CustomTrace> CustomAutoRooter<T> {
+    pub fn new() -> Self {
+        CustomAutoRooter {
+            _base: jsapi::CustomAutoRooter {
+                _vftable: &<Self as CustomAutoTraceable>::vftable,
+                _base: AutoGCRooter::new_unrooted(AutoGCRooterTag::CUSTOM),
+            },
+            data: None,
+        }
+    }
+}
+
+/// Similarly to `Trace` trait, it's used to specify tracing of various types
+/// that are used in conjunction with `CustomAutoRooter`.
+pub unsafe trait CustomTrace {
+    fn trace(&self, trc: *mut JSTracer);
+}
+
+unsafe impl CustomTrace for Vec<*mut JSObject> {
+    fn trace(&self, trc: *mut JSTracer) {
+        for obj in self {
+            let obj = obj as *const *mut _ as *mut *mut _;
+            unsafe { CallObjectRootTracer(trc, obj, c_str!("sequence<object>")); }
+        }
+    }
+}
+
+unsafe impl CustomTrace for  Vec<Value> {
+    fn trace(&self, trc: *mut JSTracer) {
+        for val in self {
+            let val = val as *const _ as *mut _;
+            unsafe { CallValueRootTracer(trc, val, c_str!("sequence<any>")); }
+        }
+    }
+}
+
+unsafe impl<T: CustomTrace> CustomTrace for Vec<Option<T>> {
+    fn trace(&self, trc: *mut JSTracer) {
+        for seq in self.iter().filter_map(|x| x.as_ref()) {
+            seq.trace(trc);
+        }
+    }
+}
+
+unsafe impl<T: CustomTrace> CustomTrace for Vec<T> {
```

Comment:
I think it makes more sense to implement CustomTrace for the following types:
* `*mut Object`
* `Value`
* `Option<T>` where `T: CustomTrace`
* `Vec<T>` where `T: CustomTrace`

This composes better than focusing on different specializations of `Vec<T>`.

Reply:
I used specialization to retain the `"sequence<any>"` and `"sequence<object>"` strings, as used in C++: https://dxr.mozilla.org/mozilla-central/source/dom/bindings/BindingUtils.h#2163.

Should I just provide `"object"` and `"any"` messages for `*mut JSObject` and `Value` and use that in `Vec<T>` where `T: CustomTrace` instead?

Reply:
Yeah, I don't think it's important to get the sequence strings here.